tests/experimental/nn/test_auto_shard.py

```diff
@@ -107,3 +105,39 @@ def test_single_run():
             input = model(input)
 
     assert input.size() == torch.Size([35, 20, 28783])
+
+
+class Branch(torch.nn.Module):
+    def __init__(self, features: int = 10):
+        super().__init__()
+        self.left = nn.Linear(in_features=features, out_features=features)
+        self.right = nn.Linear(in_features=features, out_features=features)
+
+    def forward(self, x):
+        if x.sum() > 1000:
+            return self.left(x)
+        else:
+            return self.right(x)
+
+
+class BranchedNetwork(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.net = torch.nn.ModuleList([Branch() for _ in range(10)])
+
+    def forward(self, x):
+        for module in self.net:
+            x = module(x)
+        return x
+
+
+def test_dynaimc_conditionals_auto_wrapped():
+    if torch_version() < (1, 8, 0):
+        pytest.skip("requires torch version >= 1.8.0")
+    from fairscale.experimental.nn.auto_shard import shard_model
+
+    model = BranchedNetwork()
+    sharded_model = shard_model(model, 3)
+    # TODO(ehotaj): There might be a bug in our split code because we shard the
+    # model into 10 shards even though we specify 3 shards above.
+    assert len(sharded_model) == 10
```

Comment:
Can you print out the original model and sharded model?

Reply:
Here is the full model:

```
BranchedNetwork(
  (net): ModuleList(
    (0): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
    (1): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
    (2): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
    (3): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
    (4): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
    (5): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
    (6): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
    (7): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
    (8): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
    (9): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
  )
)
```

And the branched model:

```
[GraphModule(
  (net): Module(
    (0): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
  )
), GraphModule(
  (net): Module(
    (1): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
  )
), GraphModule(
  (net): Module(
    (2): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
  )
), GraphModule(
  (net): Module(
    (3): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
  )
), GraphModule(
  (net): Module(
    (4): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
  )
), GraphModule(
  (net): Module(
    (5): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
  )
), GraphModule(
  (net): Module(
    (6): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
  )
), GraphModule(
  (net): Module(
    (7): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
  )
), GraphModule(
  (net): Module(
    (8): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
  )
), GraphModule(
  (net): Module(
    (9): Branch(
      (left): Linear(in_features=10, out_features=10, bias=True)
      (right): Linear(in_features=10, out_features=10, bias=True)
    )
  )
)]
```

Looks reasonable to me but I'm not sure why the `shard_count` is not being respected.

Reply:
> I'm not sure why the shard_count is not being respected.

BTW, this also happens without the changes in this PR, i.e. if we just use `torch.fx.symbolic_trace` directly, so I think something might be going wrong in the `_split_nodes` logic. 

(I could try looking into it in a follow up PR).